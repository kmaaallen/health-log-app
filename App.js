import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
//Redux
import { Provider } from 'react-redux';
import { store, persistor } from './src/redux/store';
import { PersistGate } from 'redux-persist/integration/react';
//Navigation
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
//Screens
import LogPage from './src/screens/LogPage';

const Stack = createStackNavigator();

export default function App() {
  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Log" component={LogPage} />
          </Stack.Navigator>
        </NavigationContainer>
      </PersistGate>
    </Provider>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
