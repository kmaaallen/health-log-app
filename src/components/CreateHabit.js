import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTheme, Button, Dialog, TextInput, Portal, Surface, HelperText } from 'react-native-paper';
//REDUX
import { connect } from 'react-redux';
import { createHabit } from '../redux/actions';
//FORM
import { useForm, Controller } from "react-hook-form";
//COMPONENTS
import { Picker } from "@react-native-picker/picker";

const styles = theme => StyleSheet.create({
    surface: {
        padding: theme.spacing.medium,
        margin: theme.spacing.medium,
    },
    pickerView: {
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 4,
        height: 100
    }
});

//TODO CHECK HOW IT LOOKS ON ANDROID
//TODO STYLE ON ALL DEVICES
//TODO ALLOW SELECTION OF MULTIPLE CATEGORIES
//TODO ALLOW UPDATE OF CATEGORIES
//LOG PAGE RE-ORDER AND FILTER HABITS
//TODO SORT OUT MOBILE MODAL OVERLAP ISSUE

export const CreateHabit = (props) => {
    const [showDialog, setShowDialog] = useState(false);

    const { control, reset, handleSubmit, watch, formState: { errors }, register } = useForm({
        mode: 'onSubmit',
    });

    const watchCategoryPicker = watch("category", false);

    const createHabit = (formData) => {
        var category;
        formData.newCategory ? category = formData.newCategory : category = formData.category;
        props.createHabit(formData.title, formData.limit, category);
        setShowDialog(false);
        reset();
    }

    return (
        <Surface elevation={4} style={styles(props.theme).surface}>
            <Button onPress={() => setShowDialog(true)}>Create new habit</Button>
            <Portal>
                <Dialog visible={showDialog} onDismiss={() => setShowDialog(false)}>
                    <Dialog.Title>Create a habit</Dialog.Title>
                    <Dialog.Content>
                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="Title"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''} // to remove uncontrolled to controlled error
                                    testID="title-input"
                                />
                            )}
                            name="title"
                            rules={{ required: true }}
                        />
                        {errors.title && <HelperText type="error" >Title is required</HelperText>}
                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="Daily limit"
                                    keyboardType="numeric"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''}
                                    testID="limit-input"
                                />
                            )}
                            name="limit"
                            rules={{ required: true, min: 1 }}
                        />
                        {errors.limit && <HelperText type="error">{errors.limit.type == 'min' ? 'Limit must be greater than zero' : 'Limit is required'}</HelperText>}

                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <View style={styles(props.theme).pickerView}>
                                    <Picker
                                        selectedValue={value}
                                        onValueChange={value => onChange(value)}
                                        itemStyle={{ fontSize: 16, height: 100 }}
                                        testID="category-input"
                                    >
                                        <Picker.Item label="Please select a category" value="" key="empty" testID="select-category" />
                                        <Picker.Item label="New category" value="New category" key="New category" testID="new-category" />
                                        {props.categories.map((item) => (<Picker.Item key={item} label={item} value={item} />))}
                                    </Picker>
                                </View>
                            )}
                            name="category"
                            rules={{ required: false }}
                        />
                        {watchCategoryPicker == 'New category' ? <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="New Category"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''}
                                    testID="new-category"
                                    placeholder="Create a new category"
                                />
                            )}
                            name="newCategory"
                            rules={{ required: false }}
                        /> : null}
                    </Dialog.Content>
                    <Dialog.Actions>
                        <Button onPress={() => setShowDialog(false)}>Cancel</Button>
                        <Button onPress={handleSubmit(createHabit)}>Ok</Button>
                    </Dialog.Actions>
                </Dialog>
            </Portal>
        </Surface >

    );
};

const mapStateToProps = (state) => ({
    categories: [...new Set((Object.values(state.habits.habits)).map((item) => item.category))].filter(item => item != "" && item != undefined).sort(),
})

const mapDispatchToProps = (dispatch) => {
    return {
        createHabit: (title, limit, category) => dispatch(createHabit({ updated: (new Date()).valueOf(), title: title, limit: limit, category: category }))
    }
}

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(CreateHabit));