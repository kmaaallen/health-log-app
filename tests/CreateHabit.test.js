import React from 'react';
import { Provider } from 'react-redux';
import { Provider as PaperProvider } from 'react-native-paper';
import { theme } from '../App';
import CreateHabit from '../src/components/CreateHabit';
// Testing
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { createTestStore } from './utils';

describe('<CreateHabit />', () => {
    let store;

    beforeEach(() => {
        store = createTestStore();
    });

    it('has 1 child', () => {
        const tree = renderer.create(<Provider store={store} ><CreateHabit theme={theme} /></Provider>).toJSON();
        expect(tree.children.length).toBe(1);
    });

    it('renders correctly', () => {
        const tree = renderer.create(<Provider store={store} ><CreateHabit theme={theme} /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders create new habit button, that when clicked shows dialog content and allows new habit to be created', async () => {
        // Paper provider component wrapper required for react-native-paper, provided at app level
        const { getByText, getByTestId, queryByText } = render(<Provider store={store}><PaperProvider theme={theme}><CreateHabit /></PaperProvider></Provider>);
        const button = getByText('Create new habit');
        // Check button is there
        expect(button).toBeTruthy();
        // Expect store to be empty at this stage
        expect(store.getState().habits).toEqual({ "habits": {} });
        // Click button
        fireEvent.press(button);
        await waitFor(() => queryByText('Create a habit'));
        // Dialog with form present
        const title = getByTestId('title-input');
        const limit = getByTestId('limit-input');
        const newCategory = getByTestId('new-category');
        const submit = getByText('Ok');
        expect(title && limit && newCategory).toBeTruthy();
        // change form values and submit form with new category
        fireEvent.changeText(title, 'My first habit');
        fireEvent.changeText(limit, '3');
        fireEvent.changeText(newCategory, 'Health');
        await waitFor(() => {
            fireEvent.press(submit);
        });
        // Expect store to be updated with one habit
        expect(Object.keys(store.getState().habits.habits).length).toBe(1);
        expect(store.getState().habits.habits[1].title).toBe('My first habit');
        expect(store.getState().habits.habits[1].limit).toBe('3');
        expect(store.getState().habits.habits[1].category).toBe('Health');
        // Expect existing category to now be present on form dialogue
        fireEvent.press(button);
        await waitFor(() => queryByText('Create a habit'));
        // expect existing category to be there now
        const existingCategory = getByTestId('category-input');
        expect(existingCategory).toBeTruthy();
    }),
        it('validates errors correctly in create new habit form', async () => {
            // Paper provider component wrapper required for react-native-paper, provided at app level
            const { getByText, getByTestId, queryByText } = render(<Provider store={store}><PaperProvider theme={theme}><CreateHabit /></PaperProvider></Provider>);
            const button = getByText('Create new habit');
            // Expect store to be empty at this stage
            expect(store.getState().habits).toEqual({ "habits": {} });
            // Click button
            fireEvent.press(button);
            await waitFor(() => queryByText('Create a habit'));
            const limit = getByTestId('limit-input');
            const submit = getByText('Ok');
            await waitFor(() => {
                fireEvent.press(submit);
            });
            // Expect store to not be updated and error messages present
            expect(Object.keys(store.getState().habits.habits).length).toBe(0);
            expect(getByText('Title is required')).toBeTruthy();
            expect(getByText('Limit is required')).toBeTruthy();
            // change limit valueto check other error message and submit form
            fireEvent.changeText(limit, '0');
            await waitFor(() => {
                fireEvent.press(submit);
            });
            expect(Object.keys(store.getState().habits.habits).length).toBe(0);
            expect(getByText('Title is required')).toBeTruthy();
            expect(getByText('Limit must be greater than zero')).toBeTruthy();
        })
});
