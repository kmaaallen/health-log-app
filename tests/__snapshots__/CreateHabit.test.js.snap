// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateHabit /> renders correctly 1`] = `
<View
  collapsable={false}
  elevation={4}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "margin": "5%",
      "padding": "5%",
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 3,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 4,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Create new habit
        </Text>
      </View>
    </View>
  </View>
</View>
`;
